docs/source/building/variants.rst

```diff
@@ -0,0 +1,911 @@
+compiler is specified in any variant. These are defined in `conda-build's
+jinja_context.py file
+<https://github.com/conda/conda-build/blob/master/conda_build/jinja_context.py>`_.
+Most of the time, users will not need to provide compilers in their variants -
+just leave them empty, and conda-build will use the defaults appropriate for
+your system.
+
+
+.. _referencing_subpackages:
+
+Referencing subpackages
+-----------------------
+
+Conda-build 2.1 brought in the ability to build multiple output packages from a
+single recipe. This is useful in cases where you have a big build that outputs a
+lot of things at once, but those things really belong in their own packages. For
+example, building gcc outputs not only gcc, but also gfortran, g++, and runtime
+libraries for gcc, gfotran and g++. Each of those should be their own package to
+make things as clean as possible. Unfortunately, if there are separate recipes
+to repack the different pieces from a larger whole package, it can be hard to
+keep them in sync. That's where variants come in. Variants, or more
+specifically, the ``pin_subpackage(name)`` function give you a way to refer to
+the subpackage with control over how tightly the subpackage version relationship
+should be in relation to other subpackages or the parent package.
+
+meta.yaml:
+
+.. code-block:: yaml
+
+   package:
+     name: subpackage_demo
+     version: 1.0
+
+   requirements:
+     run:
+       - {{ pin_subpackage('subpackage_1') }}
+       - {{ pin_subpackage('subpackage_2', pins='x.x') }}
+       - {{ pin_subpackage('subpackage_3', pins=('x.x', 'x.x')) }}
+       - {{ pin_subpackage('subpackage_4', exact=True) }}
+
+
+   outputs:
+     - name: subpackage_1
+       version: 1.0.0
+     - name: subpackage_2
+       version: 2.0.0
+     - name: subpackage_3
+       version: 3.0.0
+     - name: subpackage_4
+       version: 4.0.0
+
+Here, the parent package will have the following different runtime dependencies:
+
+* subpackage_1 >=1.0.0,<2 (default uses pins='x', pins to major version with
+  default >= current version lower bound)
+* subpackage_2 >=2.0.0,<2.1 (more stringent upper bound)
+* subpackage_3 >=3.0,<3.1 (less stringent lower bound, more stringent upper bound)
+* subpackage_4 4.0.0 h81241af (exact pinning - version plus build string)
+
+
+Compiler packages
+-----------------
+
+
+On Mac and Linux, we can and do ship gcc packages.  These will become even more
+powerful with variants, since you can specify versions of your compiler much
+more explicitly, and build against different versions (or with different flags,
+set in the compiler package's activate.d scripts) if you'd like. On Windows,
+rather than providing the actual compilers in packages, we still use the
+compilers that are installed on the system. The analogous compiler packages on
+Windows run any compiler activation scripts and set compiler flags instead of
+actually installing anything.
+
```

Comment:
I like this.  It seems like the compiler package variants and more generally host/vendor package variants are critical in the HPC environment.  Like the windows situation many of these packages would either run 'module load' commands or freeze and enforce the environment changes executed by the 'module' command.